### DISCUSSION TITLE
Databricks Pyspark notebook running for long time > 7 hours + Need recommendation on Databricks Cluster Configurations to handle

### DISCUSSION THREAD

**Discussion by maruthiservian** [moj-analytical-services/splink#1742] [Q&A]
Hi Robin,

I am currently working on utilizing Splink on Databricks to establish linking and blocking rules for the 'first_name' and 'last_name' columns across two tables. The 'ID' column serves as the cluster column with unique identifiers.

`linker.estimate_u_using_random_sampling(max_pairs=1e6)
blocking_rule_for_training = block_on(["first_name_3","last_name_3"])
#if linker.count_num_comparisons_from_blocking_rule(blocking_rule_for_training) != 0:
linker.estimate_parameters_using_expectation_maximisation(blocking_rule_for_training, estimate_without_term_frequencies=True)`

As of now, the left table comprises 467,996 rows, while the right table contains 28,209,116 rows. The Databricks configuration is as follows:
- Runtime: DBR 12.2 LTS
- Driver: Standard_DS3_v2 14 GB (4 cores)
- Workers (8): Standard_DS3_v2 112 GB (32 cores)

Despite these specifications, the code execution is taking an extended amount of time, and it neither fails nor produces an error. Code is running more than 7 hours.

I am reaching out to seek guidance and insights based on your experiences. Could you kindly provide recommendations for optimizing the Databricks cluster configuration to efficiently handle the above mentioned data?

And also I can see in splink website can process 100+ million records, could you please share any time durations regarding the same on different platforms like Databricks with PySpark, Athena etc., ?

Thank you,
Maruthi

**Comment by RobinL** [comment]
We don't have access to databricks so not sure about that, but with data of that size I'd start running your models in DuckDB.


I would recommend reading the [tutorial](https://moj-analytical-services.github.io/splink/demos/tutorials/00_Tutorial_Introduction.html). The following guides should also help you with performance:

https://moj-analytical-services.github.io/splink/topic_guides/performance/drivers_of_performance.html?h=performance
https://moj-analytical-services.github.io/splink/topic_guides/blocking/blocking_rules.html
https://moj-analytical-services.github.io/splink/demos/tutorials/03_Blocking.html?h=count#counting-the-number-of-comparisons-created-by-a-single-blocking-rule

If you share the full script you're running we might be able to spot a bug

**Reply by maruthiservian** [reply]
Dear Robin,

Sorry for late reply! Below is the code which I am trying to execute on Databricks with the above cluster configurations.
Blocking rules are running for longer time duration as below for different records size.

600k records - 25.70 minutes
1.5M records - 26.51 minutes
2.5M records - 2.47 hours
20M records - 5.43 hours

Is there a way to write the below code into better code to test splink on databricks? Please let me know and thanks in advance. I am not a pyspark coding expert and starting to learn now a days.

Below are the code details and tables information.
```sql
SELECT researcher_id_clean as researcher_id, first_name, last_name FROM TableA - 467996 records
SELECT researcher_id, first_name, last_name FROM TableB - 28455634 records
```

```python
import pandas as pd
import altair as alt
from pyspark.sql import SparkSession
from pyspark.sql.functions import col

from IPython.display import IFrame
alt.renderers.enable('html')

# Create a Spark session
spark = SparkSession.builder.appName("SPLINKExample").getOrCreate()

df_a = spark.sql("SELECT researcher_id_clean, first_name, last_name FROM TableA ")
df_b = spark.sql("SELECT researcher_id, first_name, last_name FROM TableB")

# Rename columns using withColumnRenamed
df_a = df_a.withColumnRenamed("researcher_id_clean", "researcher_id")
df_b = df_b  # No need to rename columns in df_b

# Convert PySpark DataFrames to Pandas DataFrames for compatibility with Altair
df_a_pd = df_a.toPandas()
df_b_pd = df_b.toPandas()

def prepare_data(data):
    data = data.rename(columns=lambda x: x.strip())
    data["cluster"] = data["researcher_id"].apply(lambda x: "-".join(x.split('-')[:2]))
    return data

dfs = [
    prepare_data(dataset)
    for dataset in [df_a_pd, df_b_pd]
]

display(dfs[0].head(2))
display(dfs[1].head(2))
`
-- Blocking rules

`blocking_rules = [
    "l.first_name = r.first_name",
    "l.last_name = r.last_name",
]
linker.cumulative_num_comparisons_from_blocking_rules_chart(blocking_rules)`

-- Actual comparisons with blocking rules

`import splink.spark.comparison_library as cl
import splink.spark.comparison_template_library as ctl
import splink.spark.comparison_level_library as cll


simple_model_settings = {
    **basic_settings,
    "blocking_rules_to_generate_predictions": blocking_rules,
    "comparisons": [
        cl.exact_match("first_name", term_frequency_adjustments=True),
        cl.exact_match("last_name", term_frequency_adjustments=True),
    ],
    "retain_intermediate_calculation_columns": True,
}

detailed_model_settings = {
    **basic_settings,
    "blocking_rules_to_generate_predictions": blocking_rules,
    "comparisons": [
        ctl.name_comparison("first_name", term_frequency_adjustments=True),
        ctl.name_comparison("last_name", term_frequency_adjustments=True),
        # ctl.date_comparison("date_of_birth", 
        #                     damerau_levenshtein_thresholds=[],
        #                     cast_strings_to_date=True,
        #                     invalid_dates_as_null=True,
        #                     date_format="%Y%m%d"),
        # cl.damerau_levenshtein_at_thresholds("soc_sec_id", [1, 2]),
        # cl.exact_match("street_number", term_frequency_adjustments=True),
        # cl.damerau_levenshtein_at_thresholds("postcode", [1, 2], term_frequency_adjustments=True),
        # we don't consider further location columns as they will be strongly correlated with postcode
    ],
    "retain_intermediate_calculation_columns": True,
}


linker_simple = SparkLinker(dfs, simple_model_settings, spark=spark)
linker_detailed = SparkLinker(dfs, detailed_model_settings, spark=spark)
````
Best Regards,
Maruthi

**Reply by beedrumms** [reply]
hi there! i have been using pyspark in databricks environment for a while now and noticed two things in ur code that might reduce runtime:

1)  Your blocking rules
- the use of just first_name and just last_name are likely the contributing most to your long run time. 'First-name' is very generic, even using `"l.First_Name = r.First_Name and l.Last_Name = r.Last_Name"` will likely reduce runtime but using a variation of the fields that you have in your comparison levels will reduce the comparison space. 
- this obviously wont work for your simple_model - if you must use just `"l.First_Name = r.First_Name"` and `"l.Last_Name = r.Last_Name"` for your blocking rules, you could take a sample size from your data (at least a quarter or less) and compare performance on the same sample subset. 
2) spark configurations of parallelism and partitions
- looking at your configs, you only have 8 workers, this might be to small for the amount of data you are trying to process. To support it, try setting spark.default.parallelism to around 5x the number of CPUs in your cluster. For example, a cluster w/ 8 workers, Standard_D8s_v3 32 GB memory, 8 cores (both worker and driver), the below configs works very efficiently for 1M rows.  
`spark = SparkSession.builder \
    .config("spark.databricks.service.clusterId", "####-######-wwwww") \
    .config("spark.default.parallelism", "120") \
    .config("spark.sql.shuffle.partitions", "120") \
    .getOrCreate()` 
- i found this info in the docs guide and some independent research: https://moj-analytical-services.github.io/splink/topic_guides/performance/optimising_spark.html
